### ISSUE TITLE
LibraryInstaller::getPackageBasePath() never called

### ISSUE THREAD

**Issue by akimsko** [composer/composer#4648]
At some point since alpha 11, it seems someone removed the call from getInstallPath() to getPackageBasePath().

So getPackageBasePath() is now never called in LibraryInstaller, breaking any plugins that relies on this.
Those plugins must now override getInstallPath() instead - which is fine I guess, but why leave the dead code in place?

Also the example installer code at:
https://getcomposer.org/doc/articles/custom-installers.md

No longer works...

**Comment by alcohol** [comment]
Thank you for reporting this, we will look into it. Feel free to dig into git blame/history yourself as well if you have the time available. We welcome any contributions.

**Comment by akimsko** [comment]
It was part of this commit:
https://github.com/composer/composer/commit/d98b134d

It seems getPackageBasePath() is actually still used, but only in uninstall().

There's is some discussion about it in the commit comments - but the documentation still reflects an incorrect way of doing it.

I suggest simply updating the documentation.

**Comment by Seldaek** [comment]
The docs are and were always correct.. You have to override getInstallPath
and that's it. Before we had a bug in uninstall though for those plugins,
and we fixed that. So now it behaves correctly. As I said before in the
commit comments just implement getInstallPath and if you need BC with older
composer you can duplicate getPackageBasePath from latest composer in your
plugin as it does the right thing.

**Comment by akimsko** [comment]
Ok, that's fine I guess. But am I reading different docs than you?

The example here:
https://getcomposer.org/doc/articles/custom-installers.md

Clearly shows a class overriding getPackageBasePath() only... Everything based on this example will not work.

**Comment by Seldaek** [comment]
Oh yeah sorry I missed that and that is indeed wrong. I just checked above
where it says which methods you need to implement there it's correct.

**Comment by akimsko** [comment]
Thanks :)

**Comment by spekary** [comment]
This still has a problem though, either in the docs or the code. getInstallPath is not used during the removal of a package, only getPackageBasePath, so at this point, we need to override both, and try to understand your regex replace there.

Please comment what transformation you are making in that regex replace in getPackageBasePath to make it easier for us to know whether its going to be compatible with our installation directory plans.

**Comment by Seldaek** [comment]
@spekary getPackageBasePath is just the same as getInstallPath, except it strips the package's target-dir if there is one. That's literally all it does.

Both are still in use, but getPackageBasePath defers to getInstallPath then removes the target-dir, so that implementors can override only getInstallPath and be done with it. Unfortunately that's only true for the latest composer builds, so if your plugin still has to work with older composer builds the best is to just override both and make sure getPackageBasePath forwards to getInstallPath, or if you have to support packages with a target-dir you'll want to duplicate the implementation we have at https://github.com/composer/composer/blob/master/src/Composer/Installer/LibraryInstaller.php#L151-170

**Comment by wkampmann** [comment]
Hmm. I stumbled upon this bug during deployment, and it took me quite some time to figure out what was wrong since the installer that we implemented was basically a copy/paste of the documentation. 

Since the wrong documentation has been out for so long, this change really is a breaking change for about anyone who implemented a custom installer. I think it should be treated as such. E.g. it would be advisable to add a check to the `composer diagnose` command warning about installers that override the `getPackageBasePath` method.

I was surprised to find this, since my experience with composer has been that it's extremely stable without breaking changes like this in the past, so keep up the good work :+1:.
